### ISSUE TITLE
Is PyTorch using the token embeddings ?

### ISSUE THREAD

**Issue by PonteIneptique** [hipster-philology/pandora#80]
Hi both,
I could not find any line using the token embeddings generated through Gensim in the PyTorch model. Did I miss the line or has there been a shift about using this (Or Gensim because I do see an Embedding)  ?

**Comment by PonteIneptique** [comment]
If not, we should probably avoid running this, even if this is not important for memory in the end :)

**Comment by emanjavacas** [comment]
I dont remember seeing pretrained embeddings in the keras model. Its a very
simple tweak in pytorch though so I can build it in if there is a model API
for it. I also must say that I doubt pretraining word embeddings on
pos-annotated corpora will have a benefit (given the small size of such
corpora), but itd be cool to be able to load pretrained embeddings on
bigger corpora.

**Comment by PonteIneptique** [comment]
In Keras, the pretrained embedding is loaded here : https://github.com/hipster-philology/pandora/blob/master/pandora/impl/keras/model.py#L175-L179

I am not saying we need it, I was just wondering if there was a reason it was not there :)

It seems, according to @Jean-Baptiste-Camps, that fine tuning this parameter can give up to 3% improvement on lemma accuracy. Did I understand you well JB ?

**Comment by Jean-Baptiste-Camps** [comment]
Loading pretrained embeddings from a bigger corpora would be nice indeed. It would allow me to train them on the sum of whatever Old French texts I can find, and the use them on the smaller labeled corpora I have (whose size vary between 50 k and 3 000 k).

I did notice a difference, indeed, on a small corpus (a bit less than 50 k tokens) with variation in min_token_freq_emb (basically, setting it higher allowed higher scores on train and test, which makes a lot of sense from a statistical point of view; more precisely, training was catastrophic with min_count 1, behaved better with 3, and switching from 3 to 10 increased the accuracy a lot. I would need to do more formal benchmark to see how it can be generalised. It is probably mostly a question of were to cut in the distribution, but with a floor value).

**Comment by Jean-Baptiste-Camps** [comment]
NB: on a more warning note, we should be cautious about what I just said, since I did not train the different models in the perspective of comparing the effects of this variation, so something else might be causing it.

**Comment by Jean-Baptiste-Camps** [comment]
Just to say that I am making a few experiments on embeddings with one of the tagged corpus (the largest one, 3 M tokens), and the results seem to make quite a lot of sense (as do the TSNE visualization). Here is an example (I have kept Pandora's defaults for now, with only min_count set to 10):

````python
>>> model.wv.most_similar(positive=['cheualier'])
[('cheualiers', 0.7197344899177551), ('chevalier', 0.6893125176429749), ('uallet', 0.6829389929771423), ('bacheler', 0.6448491215705872), ('vilain', 0.6261616945266724), ('prodome', 0.6113272309303284), ('preudome', 0.5974007844924927), ('vallet', 0.5890029072761536), ('uallez', 0.5840432643890381), ('ior', 0.5809659957885742)]

>>> model.wv.most_similar(positive=['dame'])
[('damoisele', 0.7411099672317505), ('pucele', 0.6799402236938477), ('amie', 0.6792483329772949), ('suer', 0.6707947254180908), ('roine', 0.6304073333740234), ('lasse', 0.6262991428375244), ('vielle', 0.6252794861793518), ('damme', 0.6095970273017883), ('fille', 0.600243866443634), ('chiere', 0.5854043960571289)]

>>> model.wv.most_similar(positive=['sui'])
[('suis', 0.8027263283729553), ('serai', 0.7987678647041321), ('fui', 0.7904558181762695), ('estoie', 0.7331922054290771), ('fusse', 0.7213727235794067), ('seroie', 0.7185894846916199), ('suj', 0.6672846078872681), ('voi', 0.6305763721466064), ('estes', 0.6271989941596985), ('aim', 0.6120427846908569)]
````
There is still a few less expected ones (form of veoir or aimer with estre, but the similarity here is the first person indicative present; vilain with chevalier?), but not much.

That makes me all the more keen on including the ability to pretrain embeddings on a larger corpus.

[embeddings.pdf](https://github.com/hipster-philology/pandora/files/1423247/embeddings.pdf)

**Comment by Jean-Baptiste-Camps** [comment]
Hmmm, looks promising at the first epoch
````
-> Epoch  1 ...
Epoch 1/1
176350/176350 [==============================] - 247s - loss: 2.8405     
::: Train Scores (lemma) :::
+       all acc: 0.8827672242699178
+       kno acc: 0.8827672242699178
+       unk acc: 0.0
::: Dev Scores (lemma) :::
+       all acc: 0.8786326336310468
+       kno acc: 0.9000936768149883
+       unk acc: 0.23055162659123055
::: Test scores (lemma) :::
+       all acc: 0.8695415589715685
+       kno acc: 0.8934253132479234
+       unk acc: 0.18548387096774194
````

**Comment by emanjavacas** [comment]
I believe this can be closed
